File: leetcode/14D Algo2/Two pointer/82. Remove Duplicates from Sorted List II.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<list>
#include<stack>
#include<unordered_set> 
#include<unordered_map> 
using namespace std;


/*
Given the head of a sorted linked list, delete all nodes that have duplicate numbers, 
leaving only distinct numbers from the original list. Return the linked list sorted as well.
*/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == nullptr || head->next == nullptr) return head;
        ListNode* dum = new ListNode();
        dum->next = head;
        ListNode* prev = dum;
        ListNode* p = head;     
        while(p != nullptr && p->next != nullptr){
            ListNode* p2 = p->next;
            if(p2->val > p->val){
                p = p->next;
                prev = prev->next;
                continue;
            }
            while(p2 != nullptr){
                if(p2->val > p->val){
                    break;
                }
                p2 = p2->next;
            }
            prev->next = p2;
            p = p2;
        }
        head = dum->next;
        delete dum;
        return head;
    }
};

File: leetcode/14D Algo2/Two pointer/82. Remove Duplicates from Sorted List II.java
import java.util.*;

public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

/*
Input: head = [1,2,3,3,4,4,5]
Output: [1,2,5]
*/

class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode dum = new ListNode(); dum.index = 0;
        dum.next = head;
        ListNode p0 = dum, p1 = head;
        while(p1 != null && p1.next != null){
            ListNode p2 = p1.next;
            //System.out.println(p0.index + " " + p1.index);
            if(p2 != null && p2.val != p1.val){
                p0 = p1;
                p1 = p2;   
                continue;
            }
            while(p2 != null && p2.val == p1.val){
                p2 = p2.next;
            }
            p1 = p2;
            p0.next = p1;
        }
        head = dum.next;
        dum.next = null;
        return head;
    }
}